luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/part_picker.rs

```diff
@@ -0,0 +1,474 @@
+        });
+        let row_height = ctx.memo(|| thumbnail_container_wh.height + ROW_VERTICAL_PADDING * 2.0);
+
+        ctx.component(scroll_view::AutoScrollViewWithCtx {
+            wh: *wh,
+            scroll_bar_width: 4.px(),
+            content: |ctx| {
+                vertical(
+                    cg_file
+                        .parts
+                        .iter()
+                        .filter(|part| part.selection_type != PartSelectionType::AlwaysOn)
+                        .map(|cg_part| {
+                            fit(
+                                FitAlign::CenterMiddle,
+                                CgPartGroup {
+                                    width: wh.width,
+                                    cg_part,
+                                    project_id,
+                                    cg_id: cg_file.id,
+                                    screen_cg,
+                                    on_event,
+                                    on_internal_event: &on_internal_event,
+                                    row_height: *row_height,
+                                    thumbnail_container_wh: *thumbnail_container_wh,
+                                    max_thumbnails_per_row: *max_thumbnails_per_row,
+                                    thumbnail_container_side_padding:
+                                        *thumbnail_container_side_padding,
+                                },
+                            )
+                        }),
+                )(Wh::new(wh.width, 100.px()), ctx);
+            },
+        });
+
+        ctx.done()
+    }
+}
+
+#[component]
+struct CgPartGroup<'a> {
+    width: Px,
+    cg_part: &'a CgPart,
+    project_id: Uuid,
+    cg_id: Uuid,
+    screen_cg: &'a ScreenCg,
+    on_event: &'a dyn Fn(Event),
+    on_internal_event: &'a dyn Fn(InternalEvent),
+    row_height: Px,
+    thumbnail_container_wh: Wh<Px>,
+    max_thumbnails_per_row: usize,
+    thumbnail_container_side_padding: Px,
+}
+impl Component for CgPartGroup<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        let Self {
+            width,
+            cg_part,
+            project_id,
+            cg_id,
+            screen_cg,
+            on_event,
+            on_internal_event,
+
+            row_height,
+            thumbnail_container_wh,
+            max_thumbnails_per_row,
+            thumbnail_container_side_padding,
+        } = self;
+
+        let no_selection = screen_cg.part(&cg_part.name).unwrap().is_not_selected();
+
+        let title_bar = render_title_bar(cg_part);
+
+        let no_selection_button = fixed(thumbnail_container_wh.width, |wh, ctx| {
+            horizontal_padding(thumbnail_container_side_padding, |wh, ctx| {
+                ctx.add(NoSelectionButton {
+                    wh,
+                    no_selection,
+                    cg_part,
+                    on_event,
+                });
+            })(wh, ctx)
```

Comment:
No need to use `(wh, ctx)`. This kind of code makes me little bit tiresome. Try not using this, then you may have more clean code I guess. I tried before and it was really cool.

Reply:
fixed with 3279e15